xapi/futures/managing-cluster-stacks/managing-cluster-stacks.md

```diff
@@ -0,0 +1,168 @@
+---
+title: Managing Cluster Stacks
+layout: design
+design_doc: true
+revision: 1
+status: proposed
+---
+
+When investigation several options for distributed and clustered
+filesystems, there have been dependencies on cluster management
+software such as O2CB for OCFS2 and corosync for GFS2.
+
+In the current Xapi toolstack there is a single global implicit
+cluster called a "Pool" which is used for: resource locking;
+"clustered" storage repositories and fault handling (in HA). In the
+long term we will allow these types of clusters to be managed
+separately or all together, depending on the sophistication of the
+admin and the complexity of their environment. We will take a small
+step in that direction by keeping the cluster management code at "arms
+length" from the Xapi Pool.join code.
+
+In
+[xcp-idl](https://github.com/xapi-project/xcp-idl)
+we will define a new API category called "Cluster" (in addition to the
+categories for
+[Xen domains](https://github.com/xapi-project/xcp-idl/blob/37c676548a53b927ac411ab51f33892a7b891fda/xen/xenops_interface.ml#L102)
+, [ballooning](https://github.com/xapi-project/xcp-idl/blob/37c676548a53b927ac411ab51f33892a7b891fda/memory/memory_interface.ml#L38)
+, [stats](https://github.com/xapi-project/xcp-idl/blob/37c676548a53b927ac411ab51f33892a7b891fda/rrd/rrd_interface.ml#L76)
+,
+[networking](https://github.com/xapi-project/xcp-idl/blob/37c676548a53b927ac411ab51f33892a7b891fda/network/network_interface.ml#L106)
+and
+[storage](https://github.com/xapi-project/xcp-idl/blob/37c676548a53b927ac411ab51f33892a7b891fda/storage/storage_interface.ml#L51)
+). These APIs will only be called by Xapi on localhost. In particular they will
+not be called across-hosts and therefore do not have to be backward compatible.
+These are "cluster plugin APIs".
+
+We will define the following APIs:
+
+- `Plugin:Membership.create`: add a host to a cluster. On exit the local host cluster software
+  will know about the new host but it may need to be restarted before the
+  change takes effect
+  - in:`cluster:string`: the uuid of the cluster
+  - in:`hostname:string`: the hostname of the management domain
+  - in:`uuid:string`: a UUID identifying the host
+  - in:`id:int`: the lowest available unique integer identifying the host
```

Comment:
In that case, maybe this would be better as a (semi-)opaque 'plugin data' field which the plugin writes back to us, rather than us guessing the rules for generating the ID?